### ISSUE TITLE
Cross-thread IO close spec is racy and unreliable

### ISSUE THREAD

**Issue by headius** [ruby/spec#828]
This spec has a race that will cause it to fail sporadically:

https://github.com/ruby/spec/blob/7526ccb1ce706a8d6548cdc46d56bf89f1518412/core/io/close_spec.rb#L47-L61

The spec attempts to detect that the thread has started reading by checking for a thread state of "stopped". However the setting of the "stopped" status and the actual blocking `read` can't be done together as a single atomic operation, allowing parallel threads to see "stopped" before the `read` has started. When this happens, the IO stream gets closed before `read` is actually called, and as a result the thread sees the normal "closed stream" error.

This race has been causing a sporadic failure on JRuby, and I would expect it to sporadically fail on other parallel-threaded implementations as well. In theory, it should also fail sporadically on CRuby, since there will be a race between the blocking thread releasing the GVL and actually making the `read` call, but the gap is much smaller and may be very difficult to trigger.

Unfortunately there's no way to fix this. It is not possible to make this test reliable with a simple thread state check, since there will always be a race. The only way I know of to write this test would be to try repeatedly to achieve some level of confidence that the expectations will pass (or won't pass) under the right conditions (which cannot be detected).

This race was discussed briefly in the original PR, but the resulting changes did not do anything to avoid it (because it is impossible to avoid).

https://github.com/ruby/spec/pull/585#pullrequestreview-90329856

I will tag this for JRuby for now.

**Comment by headius** [comment]
Example failure from JRuby:

```
IO#close raises an IOError with a clear message ERROR
Expected IOError (stream closed in another thread)
but got: IOError (closed stream)
org/jruby/RubyIO.java:3063:in `read'
org/jruby/RubyIO.java:3045:in `read'
/home/travis/build/jruby/jruby/spec/ruby/core/io/close_spec.rb:53:in `block in <main>'
/home/travis/build/jruby/jruby/spec/ruby/core/io/close_spec.rb:54:in `block in <main>'
```

This spec was tagged as a failure previously without addressing the race: https://github.com/jruby/jruby/commit/e0ee5a24d6cdb617243659d8ceae881146bbd28c

**Comment by eregon** [comment]
So for this race to happen, we need a thread switch between "set Thread status to 'sleep'" and "call read(2)", right?
That seems quite unlikely, but indeed possible
I have no good idea how to fix it either.

It seems some MRI tests uses `sleep 0.01` to workaround it: https://github.com/ruby/ruby/blob/6d77e28763ed17f75edf3b4072701b4dbd7644bb/test/ruby/test_io.rb#L3412
That's hacky of course.

Maybe we could write until the pipe is full, and then when it blocks it means the pipe is full, and let the read read more than what was written, so it will block, and then we can `close`?

I think in MRI there is some locking around IO operations, probably because an early check for closed is actually not enough, the fd could be closed in between, and the operation would process to a native call with an invalid `fd` (or potentially a new IO which reused the same `fd` number, even worse).
Does JRuby use some locking like that?
Maybe such locking actually prevents this race, if the thread status is set inside the lock, and both read and close try to acquire that lock.

**Comment by eregon** [comment]
> CRuby logically must release the GVL before making the call to `read` in order to avoid blocking all other threads.

Right. I'm actually thinking the same race can result in EBADF (or worse, using the wrong IO), i.e., if `close(2)` is called between the potential early check-if-closed in `IO#read` and the `read(2)` syscall.

I guess the only way to avoid that would be to lock around the entire read(), and around close(), but then the problem turns to how to interrupt ongoing reads/writes when calling IO#close?
Maybe via some interrupt, similar to how Thread#raise works? That would need to know which Thread has the lock, and somehow make sure that Thread doesn't return to IO#read before the interrupt (seems hard), or that the interrupt is harmless (seems easier).
Assuming that issue can be solved, then close() could simply try to acquire the lock, and keep interrupting other threads which "stole" the lock before IO#close could get it.

Seems there is actually some logic similar to that in https://github.com/jruby/jruby/blob/7a7cd8dcffebf267b890425ea92e285d1e627df6/core/src/main/java/org/jruby/RubyIO.java#L2048-L2055

The idea is with that logic, we can have this order:
* take lock
* check if closed
* change Thread status to sleep
* read(2)
* (restores status, unlock)

And then I think it's not possible to get an EBADF (IO#close would need to take the lock first), and IO#close can start just after `change Thread status to sleep`, and it will need to wait that the read(2) is interrupted, i.e., we are guaranteed close will interrupt the read(2) before it can take the lock. If the interrupt is checked before/after read(2) it shouldn't matter, we know it's from another thread, so the same error message can be used.

**Comment by headius** [comment]
The changes you propose here might help, but they would be a significant difference from current CRuby behavior which allows concurrent operations against an IO (writing while blocked on read, etc). That sort of change is out of scope for this issue, since it would not address existing these specs running on existing Ruby versions. If you want to propose this change for CRuby directly I will participate in that discussion.

For purposes of this issue, I think we should focus on ways to increase the confidence level of this spec given existing behavior.

**Comment by eregon** [comment]
@headius Could you make a PR to fix this (or fix it in JRuby's copy of ruby/spec)?
I think a limited number of tries (e.g., 10) is fine.
An extra `Thread.pass` or small `sleep(0.001)` before `read_io.close` seems also acceptable if it helps avoiding the race.

**Comment by headius** [comment]
Rather eager autoclosing there, but I have pushed a change in jruby/jruby@fa49c58e4b that appears to make this spec reliable on JRuby. It is necessarily more complex than the original, but I am willing to entertain improvements.

**Comment by eregon** [comment]
Thank you!
